pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,550 @@ package service
+				log.Error(err)
+			}),
+			statefulset.WithOnReconcileFunc(func(statefulSetList map[string][]statefulset.StatefulSet) {
+				log.Debugf("[reconcile StatefulSet] length StatefulSet[%s]: %d", r.agentName, len(statefulSetList))
+				sss, ok := statefulSetList[r.agentName]
+				if ok {
+					log.Debugf("[reconcile StatefulSet] StatefulSet[%s]: desired replica: %d, current replica: %d", r.agentName, *sss[0].Spec.Replicas, sss[0].Status.Replicas)
+					if len(sss) == 1 {
+						if *sss[0].Spec.Replicas != sss[0].Status.Replicas {
+							log.Debug("[test] skip store")
+							return
+						}
+
+						pss, ok := r.statefulSets.Load().(statefulset.StatefulSet)
+						if ok && *sss[0].Spec.Replicas < *pss.Spec.Replicas {
+							mu.Lock()
+							desiredAgentReplicas = append(desiredAgentReplicas, *sss[0].Spec.Replicas)
+							mu.Unlock()
+						}
+						r.statefulSets.Store(sss[0])
+					} else {
+						log.Infof("too many statefulset list: want 1, but %r", len(sss))
+					}
+				} else {
+					log.Infof("statefuleset not found: %s", r.agentName)
+				}
+			}),
+		)
+		if err != nil {
+			return nil, err
+		}
+	case config.REPLICASET:
+		// TODO: implment get replicaset reconciled result
+		return nil, nil
+	case config.DAEMONSET:
+		// TODO: implment get daemonset reconciled result
+		return nil, nil
+	default:
+		return nil, errors.Errorf("invalid agent resource type: %s", r.agentResourceType.String())
+	}
+
+	r.ctrl, err = k8s.New(
+		k8s.WithControllerName("rebalance storage controller"),
+		k8s.WithEnableLeaderElection(),
+		k8s.WithLeaderElectionID(r.leaderElectionID),
+		k8s.WithResourceController(job),
+		k8s.WithResourceController(rc), // statefulset controller
+		k8s.WithResourceController(pod.New(
+			pod.WithControllerName("pod discover"),
+			pod.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			pod.WithOnReconcileFunc(func(podList map[string][]pod.Pod) {
+				log.Debugf("[reconcile pod] length podList[%s]: %d", r.agentName, len(podList[r.agentName]))
+				pods, ok := podList[r.agentName]
+				if ok {
+					mu.Lock()
+					dar := make([]int32, len(desiredAgentReplicas))
+					_ = copy(dar, desiredAgentReplicas)
+					log.Debugf("[test] before creating job: desiredAgentReplicas: %v", dar)
+					mu.Unlock()
+
+					// var muPod sync.Mutex
+					// muPod.Lock()
+					// defer muPod.Unlock()
+					if len(dar) > 0 {
+						ppod, ok := r.pods.Load().([]pod.Pod)
+						log.Debugf("[test] len(pod): %d, len(prev pod): %d", len(pods), len(ppod))
+						if ok && len(pods) < len(ppod) && len(pods) == int(dar[0]) {
+							decreasedPodNames := getDecreasedPodNames(ppod, pods, r.agentNamespace)
+							jobTpl, err := r.genJobTpl()
+							if err != nil {
+								log.Infof("[recovery] error generating job template: %s", err.Error())
+								return
+							}
+							// create jobs
+							for _, name := range decreasedPodNames {
+								log.Debugf("[recovery] creating job for pod %s", name)
+								ctx := context.TODO()
```

Comment:
Should we change it to receive context from onReconcile func?
e.g)
```
func(ctx context.Context, podList map[string][]pod.Pod)
```
but this code change contains interface change...

Reply:
I think we do not have to yet.
We can ask for via authors' review.